File: 0097-interleaving-string/0097-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        rows = len(s1)
        cols = len(s2)
        
        if rows + cols != len(s3):
            return False
        
        dp = [[False] * (cols + 1) for _ in range(rows + 1)]
        
        dp[rows][cols] = True
        
        for r in range(rows, -1, -1):
            for c in range(cols, -1, -1):
                if r < rows and s1[r] == s3[r+c] and dp[r+1][c] == True:
                    dp[r][c] = True
                if c < cols and s2[c] == s3[r+c] and dp[r][c+1] == True:
                    dp[r][c] = True
                
        return dp[0][0]
        
        

File: 0097-interleaving-string/0097-interleaving-string.kt
class Solution {
    fun isInterleave(s1: String, s2: String, s3: String): Boolean {
        val m = s1.length
        val n = s2.length
        val o = s3.length
        if (s1 + s2 === s3 || s2 + s1 === s3) {
            return true
        }
        if (m + n < o || m + n > o) {
            return false
        }
        val dp = Array<IntArray>(m+1){IntArray(n+1){0}}
        dp[m][n] = 1
        for (i in m downTo 0) {
            for (j in n downTo 0) {
                if (i < m && s1[i] == s3[i+j] && dp[i+1][j] > 0) {
                    dp[i][j] = 1
                }
                if (j < n && s2[j] == s3[i+j] && dp[i][j+1] > 0) {
                    dp[i][j] = 1
                }
            }
        }
        return dp[0][0] > 0
    }
}